### ISSUE TITLE
Paths handling wierdness

### ISSUE THREAD

**Issue by ytimenkov** [gerbera/gerbera#838]
More like an ask for an opinion if it's fine to try cleaning this up or if there are any caviets I miss.

TL;DR: Use `fs::path` for everything path-related (virtual and physical paths). Convert eraly from string and late to string. Assume it uses native encoding on construction and convert via `(generic)_u8string` on output.

Don't allow escaping of slashes.

Long(er) explanation:

Gerbera started using `std::filesystem::path` internally in many places which made code around broken or inconsistent.

In most places it's related to conversion back and forth to `std::string` (e.g. for passing virtual path from `js_addCdsObject` down to `SQLStorage::addContainerChain`).

Other functions like `reduce_string` is totally unnecessary as path is normalized.

JS functions like `escapeSlash` are likely won't work or will work not as expected (or I will be really surprised). In JS code I prefer to just have `somePath.split("/")` to do the right thing and put requirements on leading / trailing slash for object properties.

**Comment by pamapa** [comment]
my idea original was to use `fs::path` for real physical path. virtual path i left to std::string. But maybe its just easier to only use `fs::path` all over. Feel free to post patches. But this patches are applied after 1.5.0 is released.

**Comment by pamapa** [comment]
One thing to consider: `fs::path` is using system slashes (on Linux `/` and on Windows `\`) where virtual path always have `/`.

**Comment by ytimenkov** [comment]
> One thing to consider: fs::path is using system slashes (on Linux / and on Windows \) where virtual path always have /.

As cppreference tells `generic_*` methods return path with forward slashes, so this should be fine.

Avoiding round-trip via `std::string` is just a convenience (although I think quite significant): since you can always tell that path is in predictable encoding and normalized so there is no question whether to escape or normalize the string (thus some performance improvement).

However I think the major issue is with correctness: the encoding conversion is done at the wrong place (at least in script): it's first turned to `path` and only then transcoding is applied and this (e.g. on windows) may blow up if this escaping function is applied.

**Comment by whyman** [comment]
> my idea original was to use `fs::path` for real physical path. virtual path i left to std::string. But maybe its just easier to only use `fs::path` all over. Feel free to post patches. But this patches are applied after 1.5.0 is released.

+1 from me.

**Comment by ytimenkov** [comment]
One other thought (while listening for various podcasts, including C++ and Unicode): Not escape the slash but rather replace it with similarly looking  the big solidus (U+29F8 ⧸), also [wikipedia](https://en.wikipedia.org/wiki/Slash_(punctuation)#Encoding).
